lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+--------------------------------------------------------------------------------
+
+-- | A sequence of blocks that starts at a given block height and slot number.
+--
+-- Values of this type model a sequence of state transitions affecting the
+-- global UTxO set:
+--
+-- @
+--    global_utxo_0 -> block_0_1 ->
+--    global_utxo_1 -> block_1_2 ->
+--    ...
+--    global_utxo_i -> block_i_j ->
+--    global_utxo_j
+-- @
+--
+-- In turn, each block contains a series of transactions that each affect the
+-- global UTxO set. The effect of applying a single block to the global UTxO
+-- set is equivalent to applying all of the contained transactions in order
+-- to the global UTxO set.
+--
+-- In general, only a subset of transactions within a block sequence will be
+-- relevant to a particular wallet, and only a subset of entries within each
+-- successive iteration of the global UTxO set will be relevant to that wallet.
+--
+-- Values of 'BlockSeq' are more convenient to generate and shrink than values
+-- of 'BlockData', but can easily be transformed into values of 'BlockData' in
+-- order to test properties of the 'applyBlocks' function.
+--
+-- Usage:
+--
+--  1. Use function 'genBlockSeq' to generate a 'BlockSeq' value.
+--  2. Use function 'blockSeqToBlockData' to convert a 'BlockSeq' value to a
+--     value of 'BlockData' suitable for use with the 'applyBlocks' function.
+--  3. Use function 'shrinkBlockSeq' to shrink to a minimal counterexample.
+--
+data BlockSeq = BlockSeq
+    { initialBlockHeight
+        :: Quantity "block" Word32
+    , initialSlotNo
+        :: SlotNo
+    , shrinkableTxSeq
+        :: ShrinkableTxSeq
+    }
+    deriving (Eq, Generic, Show)
+
+instance Arbitrary BlockSeq where
+    arbitrary = genBlockSeq
+    shrink = shrinkBlockSeq
+
+genBlockSeq :: Gen BlockSeq
+genBlockSeq = BlockSeq
+    <$> fmap toEnum (choose (0, 100))
+    <*> fmap toEnum (choose (0, 100))
+    <*> genTxSeq genUTxO genAddress
+
+shrinkBlockSeq :: BlockSeq -> [BlockSeq]
+shrinkBlockSeq = genericRoundRobinShrink
+    <@> shrinkMap toEnum fromEnum
+    <:> shrinkMap toEnum fromEnum
+    <:> shrinkTxSeq
+    <:> Nil
+
+-- | Converts a 'BlockSeq' to a value of 'BlockData' that is suitable for use
+--   with the 'applyBlocks' function.
+--
+-- This function fills every slot with exactly one block, leaving no gaps.
+--
+blockSeqToBlockData :: BlockSeq -> BlockData m addr tx state
+blockSeqToBlockData = List . blockSeqToBlockList
+  where
+    blockSeqToBlockList :: BlockSeq -> NonEmpty Block
+    blockSeqToBlockList blockSeq =
+        NE.fromList $ getZipList $ makeBlock
+            <$> ZipList (enumFrom $ blockSeq & initialBlockHeight)
+            <*> ZipList (enumFrom $ blockSeq & initialSlotNo)
+            <*> ZipList (NE.toList $ TxSeq.toTxGroupList txSeq)
+      where
+        txSeq :: TxSeq
+        txSeq = blockSeqToTxSeq blockSeq
+
+        -- Makes a block using dummy values for fields that are not relevant
```

Comment:
I am afraid that dummy values can mask a problem in case if some code suddenly relies on such a value.

Reply:
Hi @Unisay

I do understand the point you make here, and I think it's a good one, but it raises a question: what values should we use here instead?

Internally, the `applyBlocks` function does pattern match on these values, so we can't conveniently use `error`.